### PR TITLE
#156 jinja generator for basic thresholds detector

### LINKED ISSUE #156: Helper to create new detectors
Waiting for https://github.com/claranet/terraform-signalfx-detectors/issues/105 or even in complementary it could be nice to create a generator to help writing new detectors as terraform code and ease contributions.

### PR DESCRIPTION
**Pull request by xp-1000**
fix #156

### CODE REVIEW

File: scripts/templates/README.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+this will generate [detector terraform 
+resource](https://registry.terraform.io/providers/splunk-terraform/signalfx/latest/docs/resources/detector):
+
+```hcl
+resource "signalfx_detector" "real_time_requests_latency" {
+  name = format("%s %s", local.detector_name_prefix, "Database Real time requests latency")
+
+  program_text = <<-EOF
+    signal = data('read_latency_ns_p99', filter=${module.filter-tags.filter_custom})${var.real_time_requests_latency_transformation_function}
```
Comment (Shr3ps) [inline scripts/templates/README.md:105]: we don't need .publish('signal') at the end?

File: scripts/templates/detector_vars.tf.j2 [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+{% endfor -%}
+
+{% for key, rule in rules.items() -%}
+  {%- set severity = key | lower -%}
+variable "{{ id }}_threshold_{{ severity }}" {
+  description = "{{ severity | capitalize }} threshold for {{ id }} detector"
+  type        = number
+  default     = {{ rule.threshold }}
```
Comment (Shr3ps) [inline scripts/templates/detector_vars.tf.j2:51]: if threshold is not specified, j2 put "None" here, what about a if instead? (so there is no default generated)

File: scripts/templates/detector_res.tf.j2 [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        {%- set reverse_comparator = '<' %}
+      {%- else %}
+        {%- set reverse_comparator = '>' %}
+      {%- endif %}
+      {%- if '=' in rule.comparator %}
+        {%- set negative_comparator = reverse_comparator %}
+      {%- else %}
+        {%- set negative_comparator = reverse_comparator ~ '=' %}
+      {%- endif %}
+    detect(when(signal > ${var.{{ id }}_threshold_{{ severity }}})).publish('{{ severity | replace('critical', 'crit') | replace ('warning', 'warn') | upper }}')
```
Comment (Shr3ps) [inline scripts/templates/detector_res.tf.j2:28]:
```suggestion
    detect(when(signal > ${var.{{ id }}_threshold_{{ severity }}})
```

File: scripts/templates/detector_res.tf.j2 [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      {%- if 'dependency' in rule %} and when(signal {{ negative_comparator }} ${var.{{ id }}_threshold_{{ rule.dependency }}})).publish('{{ severity | replace('critical', 'crit') | replace('warning', 'warn') | upper }}')
+      {%- endif %}
```
Comment (Shr3ps) [inline scripts/templates/detector_res.tf.j2:30]:
```suggestion
      {%- else %}).publish('{{ severity | replace('critical', 'crit') | replace ('warning', 'warn') | upper }}')
```

**Review by Shr3ps — CHANGES_REQUESTED** [review]

File: scripts/templates/README.md (hunk 1 above)
Comment (xp-1000) [reply scripts/templates/README.md:105]: already fixed (commit not pushed yet)

File: scripts/templates/detector_res.tf.j2 (hunk 3 above)
Comment (xp-1000) [reply scripts/templates/detector_res.tf.j2:28]: we need this to keep the existing convention CRIT and WARN

Comment (Shr3ps) [reply scripts/templates/detector_res.tf.j2:28]: nah, this comment is paired with the next one :) (I moved the code in the else condition)

Comment (xp-1000) [reply scripts/templates/detector_res.tf.j2:28]: I think I already resolved in recently pushed commit please tell me if not.

File: scripts/templates/detector_vars.tf.j2 (hunk 2 above)
Comment (xp-1000) [reply scripts/templates/detector_vars.tf.j2:51]: yeah good catch, I think we will probably enrich the templates as things progress depending of users recurrent feedback.

File: scripts/templates/README.md (hunk 1 above)
Comment (Shr3ps) [reply scripts/templates/README.md:105]: seems not fixed?

File: scripts/templates/detector_res.tf.j2 [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+          {%- set reverse_comparator = '<' %}
+        {%- else %}
+          {%- set reverse_comparator = '>' %}
+        {%- endif %}
+        {%- if '=' in rule.comparator %}
+          {%- set negative_comparator = reverse_comparator %}
+        {%- else %}
+          {%- set negative_comparator = reverse_comparator ~ '=' %}
+        {%- endif %}
+    detect(when(signal > ${var.{{ id }}_threshold_{{ severity }}}))
```
Comment (Shr3ps) [inline scripts/templates/detector_res.tf.j2:57]: are you sure there is not too many )) at the end?

Comment (xp-1000) [reply scripts/templates/detector_res.tf.j2:57]: not sure I will do an end to end test

**Review by Shr3ps — APPROVED** [review]

File: scripts/templates/README.md (hunk 1 above)
Comment (xp-1000) [reply scripts/templates/README.md:105]: you are totally right. fixed in

File: scripts/templates/detector_res.tf.j2 (hunk 5 above)
Comment (xp-1000) [reply scripts/templates/detector_res.tf.j2]: you are totally right. fixed.

File: scripts/templates/README.md (hunk 1 above)
Comment (xp-1000) [reply scripts/templates/README.md:105]: I don't manage to reproduce, seems fixed for me ^^

**Pull request merged**